pjmedia/src/pjmedia-codec/lyra.cpp

```diff
@@ -0,0 +1,652 @@
+                                   pjmedia_codec_param *attr)
+{
+    ghc::filesystem::path model_path = lyra_cfg.model_path.ptr;
+    struct lyra_data *lyra_data = (struct lyra_data*)codec->codec_data;
+
+    pj_mutex_lock(lyra_data->mutex);
+
+    lyra_data->vad_enabled = (attr->setting.vad != 0);
+    lyra_data->plc_enabled = (attr->setting.plc != 0);
+
+    lyra_data->enc = LyraEncoder::Create(attr->info.clock_rate,
+                                         attr->info.channel_cnt,
+                                         lyra_cfg.bit_rate,
+                                         lyra_data->vad_enabled,
+                                         model_path);
+    if (lyra_data->enc == nullptr) {
+        PJ_LOG(2, (THIS_FILE, "Could not create lyra encoder"));
+        pj_mutex_unlock(lyra_data->mutex);
+        return PJMEDIA_CODEC_EFAILED;
+    }
+
+    lyra_data->dec = LyraDecoder::Create(attr->info.clock_rate,
+                                         attr->info.channel_cnt,
+                                         model_path);
+    if (lyra_data->dec == nullptr) {
+        PJ_LOG(2, (THIS_FILE, "Could not create lyra decoder"));
+        pj_mutex_unlock(lyra_data->mutex);
+        return PJMEDIA_CODEC_EFAILED;
+    }
+    lyra_data->samples_per_frame =
+                          attr->info.clock_rate / lyra_data->enc->frame_rate();
+
+    PJ_LOG(4, (THIS_FILE,"Codec opened, model_path=%s, chan_cnt=%d, bitrate=%d,"
+               " clockrate:%d, vad=%d, frame_rate=%d, samples_per_frame=%d",
+               lyra_cfg.model_path.ptr, attr->info.channel_cnt,
+               lyra_cfg.bit_rate, attr->info.clock_rate, lyra_data->vad_enabled,
+               lyra_data->enc->frame_rate(), lyra_data->samples_per_frame));
+
+    pj_mutex_unlock(lyra_data->mutex);
+    return PJ_SUCCESS;
+}
+
+static pj_status_t lyra_codec_close(pjmedia_codec *codec)
+{
+    struct lyra_data *lyra_data = (struct lyra_data*)codec->codec_data;
+
+    /* Destroy encoder*/
+    if (lyra_data->enc) {
+        lyra_data->enc.reset();
+    }
+
+    /* Destroy decoder */
+    if (lyra_data->dec) {
+        lyra_data->dec.reset();
+    }
+
+    return PJ_SUCCESS;
+}
+
+static pj_status_t lyra_codec_modify(pjmedia_codec *codec,
+                                     const pjmedia_codec_param *attr)
+{
+    struct lyra_data *lyra_data = (struct lyra_data*)codec->codec_data;
+
+    lyra_data->plc_enabled = (attr->setting.plc != 0);
+    lyra_data->vad_enabled = (attr->setting.vad != 0);
+
+    return PJ_SUCCESS;
+}
+
+static pj_status_t lyra_codec_parse(pjmedia_codec *codec,
+                                    void *pkt,
+                                    pj_size_t pkt_size,
+                                    const pj_timestamp *ts,
+                                    unsigned *frame_cnt,
+                                    pjmedia_frame frames[])
+{
+    unsigned count = 0;
+    unsigned frm_size = lyra_cfg.bit_rate / (50 * 8);

```

Comment:
Is the `lyra_cfg.bit_rate` always the same between local & remote?

Reply:
Currently yes. We can add the information on the SDP, however it would be proprietary.

Reply:
I mean does Lyra mandate that both sides *must* use the same bitrate? If not, then assuming both using same bit rate does not sound very safe, even for communication among PJSIP endpoints, as the bitrate seems to be configurable.

If Lyra does not provide packetizing APIs (haven't checked myself), IMO a safer approach would be not allowing multiple frames per packet, this is simple but perhaps not a good idea for low bitrate codecs. Another approach may be the parse just returns the whole encoded Lyra payload in the first frame and zero length data for any next frames (haven't checked if this is possible).

Reply:
The issue is that the decoder expects a single frame when decoding. It will be an issue if the payload has multiple frames.
I think using a SDP setting is the correct way, although currently, it would be proprietary.

Reply:
Using proprietary SDP is fine to me (without standard/rfc, the codec name is also kind of proprietary?). So the decoder will tell remote via its SDP (as usually SDP is about publishing decoder capability) about its expected bitrate when multiple frames per packet is used by the encoder?

Reply:
Just added the bitrate setting on the SDP.
```
a=rtpmap:96 lyra/16000
a=fmtp:96 bitrate=3200
```
The bitrate information is needed for parsing, regardless multiple frames used/not.

Reply:
I think we need to clarify in the docs:
- whether the bitrate in the SDP is requested by receiver/decoder or an information from sender/encoder,
- what will happen if SDP offer-answer use different bitrates.